### ISSUE TITLE
issue: Custom Object types are "Unpacked" instead of using the actual type

### ISSUE THREAD

**Issue by carlosbaraza** [react-hook-form/react-hook-form#8214]
### Version Number

7.30.0

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-handlesubmit-ts-v7-forked-5zu1vf?file=/src/index.tsx

### Steps to reproduce

1. Open the [codesandbox](https://codesandbox.io/s/react-hook-form-handlesubmit-ts-v7-forked-5zu1vf?file=/src/index.tsx)
2. See the errors and comments

### Expected behaviour

1. The value returned should be of the original EditorState type (or any other custom type).
2. The paths allowed should not include the properties and methods of the custom type.

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari, Edge

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Moshyfawn** [comment]
Custom object are not supported by RHF at this time. See #8030 proposal

**Comment by bluebill1049** [comment]
`unpacked` is removed in the next major version. you can test out with v8.0.0-alpha.4

**Comment by carlosbaraza** [comment]
@bluebill1049 I tested the latest alpha, but it seems like something funky is going on with TS types in that version, making it super slow in my large codebase with multiple custom objects.

@Moshyfawn @bluebill1049 how could we move forward the support for custom objects? I'm happy to contribute because this is quite critical to us integrating the library. Everything seems to work, and I thought the biggest issue was the types, because the library works just fine with my custom objects (or so it seems).

**Comment by bluebill1049** [comment]
Are you sure with v8? @carlosbaraza V8 should be a lot faster? cc @felixschorer

**Comment by bluebill1049** [comment]
Can you make sure to restart TS server as well?

**Comment by carlosbaraza** [comment]
Uhm, I have never debugged TS compiling performance. It seems like the actual `tsc` compilation is performant. But somehow, VS Code is quite slow. I restarted the language server, and even VS Code. The only change is the `v8.0.0-alpha.4`, and if I revert back, it works fine. There are many errors in my codebase when upgrading, which might have something to do with it.

![Screenshot 2022-04-20 at 11 27 56](https://user-images.githubusercontent.com/1270425/164212028-9f9c3173-765c-4328-ad2e-80ac176167b4.png)

I'll try to debug it better later and do some tracing, to see if I can pin point the culprit.

**Comment by bluebill1049** [comment]
sweet thanks would be good to know, as we are doing a lot of lazy type checks now. thanks to @felixschorer's amazing work.

**Comment by carlosbaraza** [comment]
I'm not sure I am debugging this properly, but it seems there is some nested type iteration going on and it is slowing everything down. I run a CPU profiler, but I could not still pinpoint the actual type causing the issue.

![Screenshot 2022-04-20 at 19 09 24](https://user-images.githubusercontent.com/1270425/164295443-153699bc-7fca-4be8-a1c9-0d3bd41d060f.png)

**Comment by felixschorer** [comment]
Hard to tell what is causing the slowdown. This might actually be caused by the new error type.

**Comment by carlosbaraza** [comment]
## Performance test in our project with different versions
Just did another performance test:
- `7.30.0`: Buttery smooth. Type autocomplete and jump to type definition working just fine.
- `8.0.0-alpha.1`: Slow. Type autocomplete seems to be okish, but jump to type definition takes quite some time loading, even if the type is just defined in the same file. CPU usage is about 30-40%.
- `8.0.0-alpha.4`: Unusable. Just extremely slow and CPU usage 100%. Both autocomplete and jump to type definition take ~5-10s to finish loading.

## Our project
I'm not sure how to help you reproduce this problem. Some details about our project (can't share source code unfortunately):
- We use some custom objects like DraftJS EditorState and other custom smaller interfaces with 4-5 nested.
- We have around 10 forms for a CMS with around 20 fields each, some with 3-4 WYSIWYG editors.
- The project probably has around ~1500 TSX files.
- We generate quite complex Type files using graphql codegen from a Hasura instance with ~100 tables.
- ~300 graphql queries/mutations with generated types

Basically, there is a significant amount of TS types, and if there is somehow a small overhead with the lazy optimizations, it might slow down big projects.


## Potential test
A potential test could be to performance test this:
- Generating ~10k types pointing between them to force overload
- Using some complex types like DraftJS EditorState
- Form with 20-30 fields
- Generate 10-20 forms

**Comment by felixschorer** [comment]
TS 4.5 introduced tail-call optimization for types. The new path types take advantage of this.

We should probably state TS 4.5 as a requirement for V8 then. @bluebill1049

**Comment by bluebill1049** [comment]
Will add it as a requirement and breaking change @felixschorer
